### PR TITLE
added union find method to 547 friends circles

### PR DESCRIPTION
**Pull request by gaizzzzi**

### CODE REVIEW

File: main/130_Surrounded_Regions.py [hunk 1]
```diff
@@ -30,6 +30,73 @@ def dfs_helper(x, y):
...
+                for di, dj in [[1, 0], [0, 1]]:
+                    new_i, new_j = i + di, j + dj
+                    
+                        
+                    if new_i < len(board) and new_j < len(board[0]) and board[new_i][new_j] == "O":
+                        union(new_i * len(board[0]) + new_j + 1, i * len(board[0]) + j + 1)
+                        
+        for i in range(len(board)):
+            for j in range(len(board[0])):
+                if i == 16 and j == 7:
```
Comment (sheep-sheep) [inline main/130_Surrounded_Regions.py:88]: why do we have hard coded value?

File: main/130_Surrounded_Regions.py [hunk 2]
```diff
@@ -30,6 +30,73 @@ def dfs_helper(x, y):
...
+                    pass
+                if board[i][j] == "O":
+                    if find_parent(j + i * len(board[0]) + 1) > 0:
+                        
+                        board[i][j] = "X"
+      
+
+
+
+
```
Comment (sheep-sheep) [inline main/130_Surrounded_Regions.py:100]: I think UF is a data structure, how about we make it a separate class? After the UF's initialization, we will know the cells to flip.

File: main/130_Surrounded_Regions.py [hunk 3]
```diff
@@ -30,6 +30,73 @@ def dfs_helper(x, y):
...
+                parent[parent_x] = parent_y
+            elif rank[abs(parent_x)] < rank[abs(parent_y)]:
+                parent[abs(parent_x)] = parent_y
+            elif rank[abs(parent_x)] > rank[abs(parent_y)]:
+                parent[abs(parent_y)] = parent_x
+            else:
+                parent[abs(parent_y)] = parent_x
+                rank[abs(parent_x)] += 1
+
+        # initilize the parent of edge to negative itslef
```
Comment (sheep-sheep) [inline main/130_Surrounded_Regions.py:69]: why do we want a negative value?

File: main/130_Surrounded_Regions.py (hunk 1 above)
Comment (gaizzzzi) [reply main/130_Surrounded_Regions.py:88]: didn't clean that up, it was there because of debugging

File: main/130_Surrounded_Regions.py (hunk 3 above)
Comment (gaizzzzi) [reply main/130_Surrounded_Regions.py:69]: I designed this negative value to distinguish the edge part and the center part as well as make use of original find_parent method.

**Pull request merged**